File: recorder/context.go
package recorder

import "context"

// Who a piece of work is for travels on the context rather than through every
// function between the handler and the model call.
//
// A service that threads these as parameters ends up passing three arguments
// through code that has no other reason to know about them, and the ones deepest
// down are the ones that get dropped. Set them once where a request starts.

type requestContextKey struct{}
type userContextKey struct{}
type sessionContextKey struct{}
type workspaceContextKey struct{}
type projectContextKey struct{}

// WithRequest marks a context as belonging to one end-user request.
//
// Everything recorded under this context, in this service or one it delegates
// to, groups together. That is what makes cost per unit of business work
// possible: a shortlist is produced by a request, not by a model call.
func WithRequest(ctx context.Context, request string) context.Context {
	return context.WithValue(ctx, requestContextKey{}, request)
}

// RequestFrom returns the request a context carries, if any.
func RequestFrom(ctx context.Context) string {
	request, _ := ctx.Value(requestContextKey{}).(string)
	return request
}

// User is the person a piece of work is for.
//
// The identifier is what every record carries, and it is the whole of what a
// report can group by. The name and email are what a person reading that
// report needs to know who the identifier is, and they travel differently: not
// on any record, but once per person to a directory the report is joined to.
type User struct {
	// ID is the identifier the product's own sign-in issued, exactly as that
	// sign-in names it, without a `users/` prefix or any other path. It must
	// not contain a slash: it becomes the last segment of a resource name.
	ID string

	// Name is the person's name as the product shows it. Optional.
	Name string

	// Email is the person's email address. Optional.
	Email string
}

// named reports whether there is anything to say about the person beyond
// the identifier.
func (u User) named() bool {
	return u.Name != "" || u.Email != ""
}

// WithUser marks a context as belonging to one person.
//
// Set it where the sign-in has been checked, which is the one place a product
// has the identifier and the name in hand together. Only the identifier
// reaches a record; a name given here is sent once to the directory and never
// again until it changes.
func WithUser(ctx context.Context, user User) context.Context {
	return context.WithValue(ctx, userContextKey{}, user)
}

// UserFrom returns the user a context carries, if any.
func UserFrom(ctx context.Context) User {
	user, _ := ctx.Value(userContextKey{}).(User)
	return user
}

// WithSession marks a context as belonging to one conversation.
func WithSession(ctx context.Context, session string) context.Context {
	return context.WithValue(ctx, sessionContextKey{}, session)
}

// SessionFrom returns the session a context carries, if any.
func SessionFrom(ctx context.Context) string {
	session, _ := ctx.Value(sessionContextKey{}).(string)
	return session
}

// WithWorkspace marks a context as belonging to one tenant of the organisation.
//
// The value is the bare identifier the organisation knows that tenant by, `acme` and not `organisations/dealade/workspaces/acme`, for the same reason User.ID is bare: the organisation is already configured once where the records are sent, so the library holds both halves of the name and builds it, and there is one shape of it rather than one per call site.
//
// Set it where the product's sign-in is checked, beside the user, because that is the one place a product has the tenant and the person in hand at the same time. It is deliberately not an argument at a model call site: a value a call site can choose is a value that can attribute one tenant's spend to another, and a figure that is wrong that way looks exactly like a figure that is right.
//
// A context that carries no workspace is not a gap. A product with one tenant sets nothing, its records name no workspace, and they are filed under the organisation.
func WithWorkspace(ctx context.Context, workspace string) context.Context {
	return context.WithValue(ctx, workspaceContextKey{}, workspace)
}

// WorkspaceFrom returns the workspace a context carries, if any.
func WorkspaceFrom(ctx context.Context) string {
	workspace, _ := ctx.Value(workspaceContextKey{}).(string)
	return workspace
}

// WorkspaceName joins an organisation and a bare workspace identifier into the name the platform knows that workspace by.
//
// Here rather than at each of the places that needs one, so the name a record is filed under and the name a spend decision is asked about cannot be built two different ways.
//
// An empty workspace has no name, and the caller decides what that means: a batch with no workspace is filed under the organisation, and a decision with no workspace is asked about the organisation's default. Neither is this function inventing a name for something the caller did not name.
func WorkspaceName(organisation, workspace string) string {
	if workspace == "" {
		return ""
	}
	return organisation + "/workspaces/" + workspace
}

// WithProject marks a context as belonging to one unit of work inside the workspace, such as a refresh run or a matter.
//
// Set once beside the workspace and the user, and for the same reason. A product with no such concept sets nothing.
//
// A label and never a boundary: nothing is authorised against it, and everyone who can read the workspace can read every project in it. A project is not a way to keep one reader away from another's figures — that is what the workspace is.
func WithProject(ctx context.Context, project string) context.Context {
	return context.WithValue(ctx, projectContextKey{}, project)
}

// ProjectFrom returns the project a context carries, if any.
func ProjectFrom(ctx context.Context) string {
	project, _ := ctx.Value(projectContextKey{}).(string)
	return project
}

File: python/src/agentpulse/context.py
"""Who a piece of work is for, carried on the context rather than through every function between the handler and the model call.

A service that threads these as parameters passes them through code that has no other reason to know about them, and the ones deepest down are the ones that get dropped. Set them once where a request starts, with `scope`.

They are context variables, so they follow the work across `await` and into `asyncio.to_thread`, and they do not follow it into a thread started by hand or a plain executor. `carry` takes the current values along into one of those.
"""

from __future__ import annotations

import contextlib
import contextvars
import functools
from dataclasses import dataclass
from typing import Any, Callable, Iterator, TypeVar

T = TypeVar("T")


@dataclass(frozen=True)
class User:
    """The person a piece of work is for.

    The identifier is what every record carries, and it is the whole of what a report can group by. The name and email travel differently: never on a record, but once per person to a directory the report is joined to.
    """

    # The identifier the product's own sign-in issued, bare, e.g. `8c21e0b4` rather than `users/8c21e0b4`. It must not contain a slash, because it becomes the last segment of the directory row's name.
    id: str
    name: str = ""
    email: str = ""

    def named(self) -> bool:
        return bool(self.name or self.email)


_request: contextvars.ContextVar[str] = contextvars.ContextVar("agentpulse_request", default="")
_user: contextvars.ContextVar[User | None] = contextvars.ContextVar("agentpulse_user", default=None)
_session: contextvars.ContextVar[str] = contextvars.ContextVar("agentpulse_session", default="")
_workspace: contextvars.ContextVar[str] = contextvars.ContextVar("agentpulse_workspace", default="")
_project: contextvars.ContextVar[str] = contextvars.ContextVar("agentpulse_project", default="")
_component: contextvars.ContextVar[str] = contextvars.ContextVar("agentpulse_component", default="")

_UNSET: Any = object()


@contextlib.contextmanager
def scope(
    *,
    request: str = _UNSET,
    user: User | None = _UNSET,
    session: str = _UNSET,
    workspace: str = _UNSET,
    project: str = _UNSET,
    component: str = _UNSET,
) -> Iterator[None]:
    """Marks the work inside the block as belonging to a request, a person, a tenant and so on. What is not named keeps the value it already had.

    `request` is one end-user request; everything recorded under it groups together, which is what makes cost per unit of business work possible.

    `user` is the person, set where the sign-in has been checked, which is the one place a product has the identifier and the name in hand together.

    `workspace` is the tenant of the organisation, bare, e.g. `acme`; the organisation is already known from the key. `project` is a unit of work inside it and a label only: nothing is authorised against it.

    `component` names the part of the product a model call belongs to, where the function that makes the call is not a good enough name.

    The tenant is set here and deliberately not at a model call site: a value a call site can choose is a value that can attribute one tenant's spend to another, and a figure wrong that way looks exactly like one that is right.
    """
    tokens = []
    for var, value in (
        (_request, request),
        (_user, user),
        (_session, session),
        (_workspace, workspace),
        (_project, project),
        (_component, component),
    ):
        if value is not _UNSET:
            tokens.append((var, var.set(value)))
    try:
        yield
    finally:
        for var, token in reversed(tokens):
            var.reset(token)


def current_request() -> str:
    return _request.get()


def current_user() -> User | None:
    return _user.get()


def current_session() -> str:
    return _session.get()


def current_workspace() -> str:
    return _workspace.get()


def current_project() -> str:
    return _project.get()


def current_component() -> str:
    return _component.get()


def carry(fn: Callable[..., T]) -> Callable[..., T]:
    """`fn`, bound to the request, person and tenant in force now, for handing to a thread or an executor that would otherwise run it with none of them."""
    context = contextvars.copy_context()

    @functools.wraps(fn)
    def run(*args: Any, **kwargs: Any) -> T:
        return context.run(fn, *args, **kwargs)

    return run


def workspace_name(organisation: str, workspace: str) -> str:
    """The platform's name for a tenant, e.g. `organisations/dealade/workspaces/acme`, or an empty string for no workspace."""
    if not workspace:
        return ""
    return f"{organisation}/workspaces/{workspace}"
